src/Component/src/Metadata/Resource/Factory/OperationDefaultsTrait.php

```diff
@@ -0,0 +1,142 @@
+
+        if (null === $resource->getAlias()) {
+            $resource = $resource->withAlias($resourceConfiguration->getAlias());
+        }
+
+        if (null === $resource->getApplicationName()) {
+            $resource = $resource->withApplicationName($resourceConfiguration->getApplicationName());
+        }
+
+        if (null === $resource->getName()) {
+            $resource = $resource->withName($resourceConfiguration->getName());
+        }
+
+        return $resource;
+    }
+
+    private function getOperationWithDefaults(
+        OperationRouteNameFactory $operationRouteNameFactory,
+        RegistryInterface $resourceRegistry,
+        ResourceMetadata $resource,
+        Operation $operation,
+    ): array {
+        $resourceConfiguration = $resourceRegistry->get($resource->getAlias() ?? '');
+
+        $operation = $operation->withResource($resource);
+
+        if (null === $resource->getName()) {
+            $resourceName = $resourceConfiguration->getName();
+
+            $resource = $resource->withName($resourceName);
+            $operation = $operation->withResource($resource);
+        }
+
+        if (null === $resource->getPluralName()) {
+            $resourcePluralName = $resourceConfiguration->getPluralName();
+
+            $resource = $resource->withPluralName($resourcePluralName);
+            $operation = $operation->withResource($resource);
+        }
+
+        if (null === $operation->getNormalizationContext()) {
+            $operation = $operation->withNormalizationContext($resource->getNormalizationContext());
+        }
+
+        if (null === $operation->getDenormalizationContext()) {
+            $operation = $operation->withDenormalizationContext($resource->getDenormalizationContext());
+        }
+
+        if (null === $operation->getValidationContext()) {
+            $operation = $operation->withValidationContext($resource->getValidationContext());
+        }
+
+        $operation = $operation->withResource($resource);
+
+        if (null === $operation->getRepository()) {
+            $operation = $operation->withRepository($resourceConfiguration->getServiceId('repository'));
+        }
+
+        if (null === $operation->getFormType()) {
+            $formType = $resource->getFormType() ?? $resourceConfiguration->getClass('form');
+            $operation = $operation->withFormType($formType);
+        }
+
+        $formOptions = $this->buildFormOptions($operation, $resourceConfiguration);
+        $operation = $operation->withFormOptions($formOptions);
+
+        if ($operation instanceof HttpOperation) {
+            if (null === $operation->getRoutePrefix()) {
+                $operation = $operation->withRoutePrefix($resource->getRoutePrefix() ?? null);
+            }
+
+            if (null === $operation->getTwigContextFactory()) {
+                $operation = $operation->withTwigContextFactory('sylius.twig.context.factory.default');
```

Comment:
Unrelated to current PR, but why metadata factory is aware of the service ID? Should be handled via tagging/configuration/interfaces instead.